### ISSUE TITLE
Customizable Nullability for UDFs and UDAFs

### ISSUE THREAD

**Issue by jayzhan211** [apache/datafusion#11274]
### Is your feature request related to a problem or challenge?

Follow on #11093 and also there is another issue #11256  expect this

### Describe the solution you'd like

There are some functions always returns non-null result, like `count`, `array_agg`, `min/max` and more.
We should able to define nullability for each UDAF

**Comment by findepi** [comment]
To address this we could extend the pattern here
https://github.com/apache/datafusion/blob/4bc322819f28695b6b9593d5bbaea775e7d798b6/datafusion/expr/src/aggregate_function.rs#L84-L88

ie now we have `return_type(input_types, nullability) -> return_type`
we could have `return_type(input_types, nullability) -> (return_type, nullability)`

this is an API for builtin functions, we would need similar API for UDAFs
https://github.com/apache/datafusion/blob/5f02c8a0658d9ed012ca4ca91f0417a967fe094a/datafusion/expr/src/udaf.rs#L330

Obviously, this would be a breaking change: both method signature and return type would change.
We can avoid the breaking change by making nullability a separate method, but it would be good to first determine the ideal end-state, and only then think how to get there.

**Comment by findepi** [comment]
Just thinking more about this. SQL spec

> If no row qualifies, then the result of COUNT is 0 (zero), and the result of any other aggregate function is the null value.

and this is also about array_agg.

So it might be that the `count` is the only aggregate function that really returns non-null values, and maybe we don't need to complicate UDAF API for this.

**Comment by jayzhan211** [comment]
I agree that we don't need non-null for `array_agg`, `min/max`, but we may need it for count variant like `approx_count_distinct` or other user-defined function that expect non-null, so able to define nullability still seems like a good idea.

**Comment by jayzhan211** [comment]
Another thing to note that is the `nullability` of the function is different from the final result but the information for optimizer.

We can still tell optimizer that the function is always non-null and do the optimization based on this information. We can still get `null` as the final result as no row qualifies, so we get the empty plan which is null value.

https://github.com/apache/datafusion/pull/11093#issuecomment-2188857213

**Comment by Jefffrey** [comment]
Seems resolved by #11989, see available methods:

https://github.com/apache/datafusion/blob/c6f73636b771ddb827c2c0b4c5c690707867e9a0/datafusion/expr/src/udaf.rs#L508-L537
